src/api/assets/assetsHooks.ts

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2018 - 2022 The Alephium Authors
+This file is part of the alephium project.
+
+The library is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+The library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { ALPH } from '@alephium/token-list'
+import { useQuery } from '@tanstack/react-query'
+import { flatMap } from 'lodash'
+
+import { useQueriesData } from '@/hooks/useQueriesData'
+import { UnverifiedNFTMetadataWithFile, VerifiedFungibleTokenMetadata } from '@/types/assets'
+
+import client from '../client'
+import { assetsQueries } from './assetsApi'
+
+export const useAssetMetadata = (assetId: string) => {
+  const isAlph = assetId === ALPH.id
+
+  const { data: allVerifiedTokensMetadata } = useQuery({
+    ...assetsQueries.metadata.allVerifiedTokens(client.networkType),
+    enabled: !isAlph,
+    staleTime: Infinity
+  })
+  const verifiedTokenMetadata = allVerifiedTokensMetadata?.find((m) => m.id === assetId)
+
+  const { data: assetBaseRaw } = useQuery({
+    ...assetsQueries.type.details(assetId),
+    enabled: !isAlph && !verifiedTokenMetadata,
+    staleTime: Infinity
+  })
+  const assetType = assetBaseRaw?.type
+
+  const { data: unverifiedFungibleTokenMetadata } = useQuery({
+    ...assetsQueries.metadata.unverifiedFungibleToken(assetId),
+    enabled: !isAlph && !verifiedTokenMetadata && assetType === 'fungible',
+    staleTime: Infinity
+  })
+
+  const { data: unverifiedNFTMetadata } = useQuery({
+    ...assetsQueries.metadata.unverifiedNFT(assetId),
+    enabled: !isAlph && !verifiedTokenMetadata && assetType === 'non-fungible',
+    staleTime: Infinity
+  })
+
+  const { data: nftData } = useQuery({
+    ...assetsQueries.nftFile.detail(assetId, unverifiedNFTMetadata?.tokenUri ?? ''),
+    enabled: assetType === 'non-fungible' && !!unverifiedNFTMetadata?.tokenUri,
+    staleTime: Infinity
+  })
+
+  const unverifiedNFTMetadataWithFile: UnverifiedNFTMetadataWithFile | undefined = unverifiedNFTMetadata
+    ? { ...unverifiedNFTMetadata, file: nftData }
+    : undefined
+
+  if (isAlph) return { ...ALPH, type: 'fungible', verified: true } as VerifiedFungibleTokenMetadata
+
+  return (
+    verifiedTokenMetadata ||
+    unverifiedNFTMetadataWithFile ||
+    unverifiedFungibleTokenMetadata || { id: assetId, type: 'unknown', verified: false }
+  )
+}
```

Comment:
First answer (need to dig in this deeper):

> This makes me think: Should we add the data loadingstatus of the previous query to the next one in its enabled dependency flag to avoid unnecessary calls?

Yes, that's definitely something I want to add.

> The above will happen for every mount of an Amount or AssetLogo component until the assetsQueries.metadata.allVerifiedTokens returns a value... this might amount to a few dozen of needless queries.

As you mentioned it, if a query has been executed (or is being executed I believe) somewhere, and we're trying to call the same query identified by the same tag, the call shouldn't be broadcasted again. 

> Performance

Thanks for raising this topic. We need to have a conversation about optimisation. Not only code optimisation, but a more holistic UX optimisation. Regarding the code, it's good to identify the low hanging fruits (ie. obvious iterations that are not needed etc.), and your feedback will definitely help me doing the first optimisation pass. But we should be mindful and avoid getting down the rabbit hole of over-optimisation.

As of today, and before my first optimisation pass, the initial loading of a big address (1DrDy) on a pretty old computer (7 yo mac) feels much faster than loading the same address in the DW. This of course is caused mostly by loading state management and query retry strategy (which we talked about improving in the DW), but what I mean is: optimising UX > over-optimising code. And IMO, simple (readable) code > over-optimised complex code. 

Some thousand of simple loops (not involving complex logic) shouldn't have a big impact on performances. I'd argue that we should monitor the execution time rather than the number of calls.

Anyway, of course if we find easy ways to optimise, we should do it. We should just try to strike the right balance between optimisation / readability / UX :)

Reply:
I agree with you. 

> Or figure out how to use a Map as the cached object instead of an Array

If we find out how to transform the cache to a map we could benefit from not having to do a full array lookup every time a component re-renders by simply getting it by key. That won't complexify the code, if anything, it'll make it simpler.

Reply:
As for the DW, the we haven't given love to the loading speed as of yet. It's not fair to compare it without first spending some time to optimize things. Our priority was to ship tokens/nft metadata asap and focus on mobile wallet. I would like to have the chance to optimize the loading speed before drawing conclusions :)

Reply:
>  I would like to have the chance to optimize the loading speed before drawing conclusions

Of course ! I *know* that the experience will be top-notch once we take a look at our loading strategy. That's what I tried to say clumsily when mentioning the loading state mgmt that we plan to improve. My point was just : we should in general aim to optimise the UX first and code second, and not the other way around (which again - I'm not saying that we do that in the DW. UX in the DW is already perfect for 99% of our users with few addresses).